Resolve source maps in error stack traces

jnicklas opened the pull request:
Closes #431

Use the error-stack-parser and get-source packages to extract the source location from stack traces and transform these sources. This gives us vastly improved stack traces.

Also adds support for filtering the stack trace to hide lines from bigtest itself, as well as dependencies from node_modules. The resulting stack traces are very clean. In the future we can make these configurable via options to the CLI.

Example:

```
⨯ [step]       Failing Test -> child -> child second step
|    ERROR: moo
|      test/fixtures/failing.test.ts:14:54 @ _callee2$
```

jnicklas committed:
Add support for variables in GraphQL queries

jnicklas committed:
Resolve source maps in error stack traces

Closes #431

Use the error-stack-parser and get-source packages to extract the source location from stack traces and transform these sources. This gives us vastly improved stack traces.

Also adds support for filtering the stack trace to hide lines from bigtest itself, as well as dependencies from node_modules. The resulting stack traces are very clean. In the future we can make these configurable via options to the CLI.

Example:

```
⨯ [step]       Failing Test -> child -> child second step
|    ERROR: moo
|      test/fixtures/failing.test.ts:14:54 @ _callee2$
```

cowboyd commented on packages/agent/app/lane.ts:
```diff
@@ -71,7 +64,7 @@ function *runLaneSegment(testRunId: string, agent: Agent, test: TestImplementati
         } catch(error) {
           console.error('[agent] assertion failed', assertion, error);
-          agent.send({ testRunId, type: 'assertion:result', status: 'failed', error: serializeError(error), path: assertionPath });
+          agent.send({ testRunId, type: 'assertion:result', status: 'failed', error: yield serializeError(error), path: assertionPath });
```
What do we do if an error occurs in error serialization? I think we need to make sure that this is a very safe thing, because as-is, it could blow up the entire test run.

jnicklas replied:
Yes, you're right, we should be careful here!

cowboyd commented on packages/agent/app/serialize-error.ts:
```diff
@@ -0,0 +1,40 @@
+  };
+}
+
+export function *serializeError(error: Error): Operation<ErrorDetails> {
```
Even though our type is `Error` here, I think we can enter this function with values that are not, in fact, errors because the default type of a catch block is `any`. While it should never happen that someone does something like `throw 73;` inside their test cases, I think we should abide by the heuristic: "if something can be thrown, it will be"

jnicklas replied:
Probably true, but what is a good way of disambiguating here? Should we just check `typeof`?

cowboyd replied:
Maybe we could duck type on  a `stack` property or whatever is used by `ErrorStackParser.parse()`?

dagda1 replied:
You could have a type guard

```typescript
function isError(error: Error | any): error is Error {
  return error instanceof Error;
}
```

That is where the line below that was queried in one of my previous PRs comes in for `BundlerError`

```typescript
Object.setPrototypeOf(this, new.target.prototype);
```

All custom errors should do this IMO.  It should be added to `AmbiguousElementError` and friends

type guards also type narrow so you could  have it

`*serializeError(error: unknown)`

If the type guard returns true then the type is narrowed to the Error class

jnicklas replied:
@dagda1 I think pretty much all the errors we use inherit from `error` via `class SomeError extends Error`, this should satisfy the `instanceof` check, no prototype trickery necessary.

dagda1 replied:
@jnicklas the reason I mention it is [this](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#example).

I've always believed it was necessary as according to the link, the prototype chain is broken and needs to be restored.

Definitely not saying you must do this

cowboyd commented on packages/server/src/manifest-server.ts:
```diff
@@ -7,12 +7,17 @@ interface ManifestServerOptions {
-  app.raw.use(staticMiddleware(options.dir));
+  app.raw.use(staticMiddleware(options.dir, {
+    setHeaders(res) {
+      res.setHeader('Access-Control-Allow-Origin', `http://localhost:${options.proxyPort}`);
```
Is this required so that the stacktrace resolver can fetch the sourcemap?

jnicklas replied:
Exactly.

cowboyd commented on packages/cli/src/query.ts:
```diff
@@ -1,8 +1,8 @@
 export function run() {
   return `
-    subscription {
+    subscription($showInternal: Boolean! = true, $showDependencies: Boolean! = true) {
```
while `showInternal` makes sense as the parameter to the `stack()` property on `ErrorDetails`, it threw me for a loop when I was reading it at the subscription level, can we call these variables `stacktraceShowInternal` and `stacktraceShowDependencies` ?

cowboyd reviewed (commented):
A few suggestions, but I love it how BigTest does the heavy lifting here to resolve stack traces and provide a convenient way of filtering them over GraphQL. This provides the best of both worlds, where building the best error message is easy, and working with complex stacks is possible.

taras reviewed (commented):
Does it make sense to remove the invocation function because it's not very helpful at the moment: `@ _callee2$`

jnicklas commented:
@taras yeah, I've been unable to get better information for the function name, but I still think that it could be useful sometimes, e.g. if the error is thrown in a helper method or something. Maybe we could make this nicer to read through using terminal colors, we could de-emphasize the function name, but still render it.

taras commented:
@jnicklas is it possible to get the actual code at that line?

jnicklas commented:
@taras yes it is, `getSource` makes it pretty simple to get the code at the location.

jnicklas committed:
Upgrade chromedriver

jnicklas committed:
Defensive coding in serializeError

Make sure errors when serializing error don't propagate upwards. If an error stack is not able to be resolved it should not cause any other problems.

jnicklas committed:
Rename query variables for clarity

jnicklas committed:
Optionally show source code in error stack

This extracts the source line at the location of the error and makes it accessible to the client. In the CLI we can optionally show this, but it is disabled by default since it's a bit verbose.

jnicklas commented:
@taras I added support for getting the source at the error location, but it is disabled by default. We should add some parameters to the CLI to show these, but I didn't want to bloat this PR up even further.

taras commented on packages/cli/src/query.ts:
```diff
@@ -11,10 +11,18 @@ export function run() {
-          lineNumber
-          columnNumber
-          stack
+          stack(showInternal: $showInternalStackTrace, showDependencies: $showDependenciesStackTrace) {
+            name
+            fileName
+            code @include(if: $showStackTraceCode)
+            line
+            column
+            source {
+              fileName
+              line
+              column
+            }
+          }
```
WOW

taras reviewed (approved):
👍
